feature: Live Activity

bastiaanv opened the pull request:
This PR is an initial/basic version of this Live Activity, without any feature which are spoken of in the Zulipchat.

## Look & Feel:

### Lock Screen (all iPhone's):
| History only                                                                                 | History and predictive line                                                                  | Live activity stale (8h limit i.e.)                                                                 |
|----------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------|
| ![IMG_2536](https://github.com/user-attachments/assets/f09c4e87-9bcc-4721-ad02-23de875e522a) | ![IMG_2534](https://github.com/user-attachments/assets/47b11cac-688e-4f09-9725-044a76df0ee3) | ![IMG_2535](https://github.com/user-attachments/assets/d3fde464-85e4-4897-94da-1a313ce5f068) |

### Dynamic Island (iPhone 14 pro max and higher):

> Screenshots are from a iPhone 15 pro max simulator

| Dynamic Island minimal                                                                                                                                | Dynamic Island expanded                                                                                                                               |
|-------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------|
| ![Simulator Screenshot - iPhone 15 Pro Max - 2024-07-15 at 12 49 36](https://github.com/user-attachments/assets/e1fdc792-14b9-4707-83fe-d8b04d40186c) | ![Simulator Screenshot - iPhone 15 Pro Max - 2024-07-15 at 12 51 14](https://github.com/user-attachments/assets/2ac1f2ac-c553-457d-8b09-4f2891b64c97) |

### Settings

> Can be found at: Settings -> Alert management -> Live Activity

| Root settings                                                                                | Bottom row settings                                                                          | Available datapoints for bottom row                                                          |
|----------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------|
| ![IMG_2537](https://github.com/user-attachments/assets/a17075df-21e4-4a58-89a9-0a29705ace75) | ![IMG_2538](https://github.com/user-attachments/assets/f023249d-3089-4a39-b26c-3a30cb051636) | ![IMG_2539](https://github.com/user-attachments/assets/70da3268-5331-4e84-9616-e09fe29f0f43) |

## Things to know:

- The bottom row has a max of 4 items, but less items can be configured for small screen sizes (like the iPhone 13 mini)
- The bottom row can be configured to the user's wishes. Swap items around, different order etc. The default is: `CurrentBG`, `IOB`, `COB`, `UpdatedAt` (like in the screenshots)
- The following events trigger a re-render of the Live Activity:
  - LoopDataManager -> `GlucoseStore.glucoseSamplesDidChange`
  - LoopDataManager -> `CarbStore.carbEntriesDidChange`
  - LoopDataManager -> `dosing changing`
  - NotificationCenter -> `UIApplication.willEnterForegroundNotification`

bastiaanv committed:
start: Start live activity project

bastiaanv committed:
style: Rework live activity

bastiaanv committed:
wip: Wip live activity

bastiaanv committed:
wip: Update live activity

bastiaanv committed:
wip: Continue live activity. TODO's expanded view di, localization

bastiaanv committed:
Merge branch 'LoopKit:dev' into feat/live-activity

bastiaanv committed:
fix: Allow configurable chart limits

bastiaanv committed:
feat: Add stale state, minor chart fixes & fix delta

bastiaanv committed:
fix: Minor fixes

bastiaanv committed:
fix: Minor hotfixes

bastiaanv committed:
chore: Cleanup

bastiaanv committed:
style: Minor styling fixes

bastiaanv committed:
fix: fix delta calculations & optimize booting sequence of LA

bastiaanv committed:
Merge pull request #2 from LoopKit/dev

Bump version to 3.5.0 to signify dev branch

bastiaanv committed:
fix: minor fix & cleanup

bastiaanv committed:
style: Fix dynamic island expanded view

A9qx commented:
Love the draft/prototype! Would love to see this make it into the project down the line I think it looks amazing!

marionbarker commented:
Built onto my real phone. Thanks so much for making this available. 👍

bastienmoulia commented:
Awesome work !
Is it possible to change the y-axis of the graph in order to be more center on the data like in the app?
![image](https://github.com/user-attachments/assets/48c6984f-685b-482a-8c36-22dddc771178)
![image](https://github.com/user-attachments/assets/c77363cd-c833-4244-be00-b0d66b1c93aa)

marionbarker commented:
I have been using this for several days - I really enjoy having it attached to Loop.

### Bug report:
* Dynamic Island expanded (long press on dynamic island)
   - truncates the display when glucose is in mg/dL and above 99 mg/dL
   - graphic below with 24 hour display disabled (top) and enabled (bottom)

![loop-pr-2191-dynamic-island-expanded](https://github.com/user-attachments/assets/d3ed86c4-d86b-4763-a021-a4cc08814c71)

### Feature requests: 
* see the target range shown on the Look Screen Widget graph
* see if an override is currently enabled, definitely added to target display on graph, possibly with the override name displayed

bastiaanv committed:
fix: Fix truncating text in expanded Dynamic Island

bastiaanv committed:
feat: Add glucose targets & preset targets to Live Activity

marionbarker commented:
Nice upgrade except the override appears confused.

* This override was active from 11:05 to 1:05 pm (with 2 hours duration), but it shows as being currently active (1:45 pm).
* I see the same pattern when long-pressing the dynamic island.

![old-override-shows](https://github.com/user-attachments/assets/bb861f65-2247-475f-90c4-0000bdbf8c58)

* I added a new override at 1:47 pm that has a duration of 3 hours.
* I will report back in 20 minutes or so. Long enough to see if the start point is correct
* And then I will turn it off and see what the plots look like.

marionbarker commented:
The walking override started at 1:47 with a duration of 3 hours and was cancelled 19 minutes later.
* upper plot: While override is active. The bad news is the start point for the plot for both lock screen and dynamic island expanded plot is "now" instead of when override started.
* lower plot: After disabling the override. The good news is that plotting a cancelled override was not repeated - perhaps that was related to the new build, or perhaps because it was cancelled instead of allowed to run the full duration.

![override-status](https://github.com/user-attachments/assets/8a3215ab-99eb-4ef5-ada3-5c306674ebaf)

bastienmoulia commented:
Thanks for the blue area!

Same here for the bug with an override label shown even none were triggered
![IMG_B15E4FD11EDD-1](https://github.com/user-attachments/assets/abc58739-6e4d-46fc-9d30-bb1e454eb064)

marionbarker commented:
I did a bit more testing. 
* If an override is active, it is shown as starting "now"
* If an override expires after the planned duration, it continues to be show starting "now"
* If an override is cancelled early, no extraneous override is displayed

bastiaanv committed:
fix: Fix preMeal override & remove glucoseTarget while override is active

bastienmoulia commented:
In the last update, I didn't see the target area for the next hour by default:
![IMG_598D1AB784AD-1](https://github.com/user-attachments/assets/cdd92914-f83e-4c51-9b5e-9d506280bd46)

If I activate the premeal, then remove it, it's ok.

bastiaanv committed:
wip

bastiaanv committed:
merge

bastiaanv committed:
fix: Restruct preset code

bastiaanv commented:
Aight, it has been a crazy few days, but I made several fixes and restructured the code a bit. This should fix most of the weird behaviour found by @bastienmoulia and @marionbarker.

@marionbarker regarding the extraneous override. With this new version, I made it so that the preMeal override is "preferred" above the scheduled overrides, since that is only active for a maximum between now and 1hour. Do you agree with this? I dont think there is enough space to show both texts, tho it is possible to show both in the graph

bastiaanv commented:
Other question, do we want to keep the upper and lower limit in the settings (and with that the green, orange and red coloring in the graph)?

bastienmoulia commented:
> Other question, do we want to keep the upper and lower limit in the settings (and with that the green, orange and red coloring in the graph)?

Personally I prefer to not have multiple color because it causing more stress when it's orange or red. So no limit.

marionbarker commented:
I have no strong opinion on the color. But with the ability to modify the glucose levels at which colors change allows people who want colors to have them and folks who do not want colors to turn them off.

I just set mine to 40 and 400 mg/dL to give it a try.

bastiaanv committed:
feat: Allow to disable colored chart

bastiaanv committed:
fix: Minor fixes for BG coloring

bastiaanv commented:
I added the setting to enable or disable the colouring based on the BG

| Chart view without colouring                                                                                                                          | New setting (with BG colouring enabled)                                                                                                                                          |
|-------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------|
| ![Simulator Screenshot - iPhone 15 Pro Max - 2024-08-01 at 20 43 44](https://github.com/user-attachments/assets/1822d6fc-3be5-4204-8888-b9740b95085a) | ![Simulator Screenshot - iPhone 15 Pro Max - 2024-08-01 at 20 45 36](https://github.com/user-attachments/assets/5bd83277-4c9a-49da-81eb-de087e3a9aa8) |

bastiaanv committed:
feat: Add carbohydrate history

bastiaanv committed:
feat: Add small mode & minor rework bottom row

bastiaanv committed:
chore: Minor code cleanup

bastiaanv committed:
revert: Revert unrelated commit for LA

marionbarker commented:
## Bug reports:
1. I tested with the latest and I am still having the Live Activity show overrides that were allowed to end at the planned duration. It does clear by setting and unsetting any override on the Loop app.
2. Adjust the values for BG coloring, there does not appear to be a way to save these values

## Comments
I would prefer if there was an explicit Save button to save new settings on the Live Activities screen with the Back button leaving settings alone.

## Successes
The things that work as expected:
* Choose Large or Small
* Available for Large or Small:
   * Add / Delete / Rearrange items for bottom row
* Available for Large only:
   * Enable / Disable Prediction Line
   * Enable / Disable Use BG coloring
   * Adjust the values for the BG coloring (see bug report)

rfmurphy81 commented:
Dumb question, but could anyone point me to documentation that could show me how to try out this branch? Downloading the ZIP from https://github.com/bastiaanv/Loop/tree/feat/live-activity doesn't seem to be buildable in Xcode but perhaps I am missing a step.

bastienmoulia commented:
> Dumb question, but could anyone point me to documentation that could show me how to try out this branch? Downloading the ZIP from https://github.com/bastiaanv/Loop/tree/feat/live-activity doesn't seem to be buildable in Xcode but perhaps I am missing a step.

I don't know how to do this with Xcode, but with the build from github I change the .gitmodules file like this :
https://github.com/bastienmoulia/LoopWorkspace/blob/dev/.gitmodules
And when there is an update I run `git submodule update --remote` in the LoopWorkspace before pushing it.

bastiaanv commented:
There are several ways to do this:
1. Open the terminal in the Loop folder. Do 'git remote set-url origin https//github.com/bastiaanv/loop', then do 'git pull', then do 'git checkout feat/live-activity'

2. Or download the patch file from https://github.com/loopandlearn/customization, open the terminal in the Loop folder, then do 'git apply <path-to-patch-file>

Please note, that I typed this on my phone, so there might be typos in my commands

partcoffee commented:
## Bug report
When using Nightscout as CGM, it appears that the data will always be reported as stale. For additional context on this issue, Nightscout is reading from LibreLinkUp, and thus reporting blood glucose values on a 1-minute interval, which may affect how this is handled.

Using v3.4.1 with the [latest customization](https://github.com/loopandlearn/customization/pull/41) as of today.

bastiaanv commented:
Hi @partcoffee ,

This behaviour is correct. The Live Activity doesnt keep iOS from suspending the app (also called Heartbeath). This also means that Loop isn't doing anything while the app is suspended. Jumping to conclusion, I assume you use Loop as a following app (just watching the BG values from NightScout).

There are a few ways to correct this:
1. Connect a BLE-enabled pump to Loop
2. Customize Loop to include background sounds (like [LoopFollow](https://github.com/loopandlearn/LoopFollow/blob/main/LoopFollow/Helpers/BackgroundTaskAudio.swift)) or use Location services to keep Loop from being suspended
3. Or switch to something like [xDrip4ios](https://github.com/JohanDegraeve/xdripswift) which is a dedicated app to follow BG values from different sources like [NightScout](https://xdrip4ios.readthedocs.io/en/latest/connect/follower/) or [LibreLinkUp](https://github.com/JohanDegraeve/xdripswift/tree/master/xdrip/Managers/LibreLinkUp)

partcoffee commented:
> Hi @partcoffee ,
> 
> 
> 
> This behaviour is correct. The Live Activity doesnt keep iOS from suspending the app (also called Heartbeath). This also means that Loop isn't doing anything while the app is suspended. Jumping to conclusion, I assume you use Loop as a following app (just watching the BG values from NightScout).
> 
> 
> 
> There are a few ways to correct this:
> 
> 1. Connect a BLE-enabled pump to Loop
> 
> 2. Customize Loop to include background sounds (like [LoopFollow](https://github.com/loopandlearn/LoopFollow/blob/main/LoopFollow/Helpers/BackgroundTaskAudio.swift)) or use Location services to keep Loop from being suspended
> 
> 3. Or switch to something like [xDrip4ios](https://github.com/JohanDegraeve/xdripswift) which is a dedicated app to follow BG values from different sources like [NightScout](https://xdrip4ios.readthedocs.io/en/latest/connect/follower/) or [LibreLinkUp](https://github.com/JohanDegraeve/xdripswift/tree/master/xdrip/Managers/LibreLinkUp)

Thanks for the clarification, but I'm using Nightscout as a CGM with Omnipod Dash, which functions as a heartbeat as well from my understanding. Loop is running its normal closed loop activities in the background like bolusing and suspending basal based on the readings from Nightscout.

My CGM is not compatible unfortunately, which is why Nightscout is being used.

The Live Activity is reporting the data as stale immediately, and not after a few minutes, when the app in theory still would be "awake".

bastiaanv commented:
@partcoffee My bad. Thank you for the clarification!

I think you are spot on with the 1 minute time interval. The LA has a limit for the amount of data I can send to the widget, too much data leads to the Widget always being stale. So I can limit the amount of samples being send to the widget even further, but that would limit the usages of the LA. You could try to limit the glucoseSamples yourself for the LA maybe to 70: https://github.com/bastiaanv/Loop/blob/feat/live-activity/Loop/Managers/Live%20Activity/GlucoseActivityManager.swift#L183

But the best solution would be to limit the amount of Glucose samples in Nightscout. Maybe the import script you are running allows for something like that?

bastiaanv committed:
fix: Process feedback

bastiaanv commented:
@marionbarker I fixed your reported bugs! I also made that when you open Loop and the LA isn't showing but is activated, that Loop will recreate the LA, without the need to re-enable LA or kill Loop

partcoffee commented:
@bastiaanv Thanks for looking in to it! I'll do some tinkering with it myself, I know it's a very niche issue.

I think the best approach would be to alter Nightscout, as you suggested. Loop runs a lot better with 5 minute intervals.

bastiaanv commented:
Good luck! If you need help, you can always contact me via DM on Zulipchat

bastiaanv committed:
style: Add save button

bastiaanv committed:
style: Minor bg coloring fix

bastiaanv commented:
@marionbarker Update the PR to include a save button and minor bg coloring fix

marionbarker commented:
Not published yet in CustomizationSelect. 
Testing prior to publishing.

## Test result 1
The stale override that expired at end of duration is fixed.

## Save is almost ok
The save button is there but does not seem to work quite right.
1. If I go to the Live Activities screen
   * make a modification 
   * save it and then use the Back button and then return to the Live Activities screen - it shows previous value (before I hit save)
2. If I subsequently view the LA widget on the Lock Screen, the widget does show the update
3. If I then go back to the Live Activities display, the modifications shows as a requested in Step 1

I rebooted the phone and got the same behavior.

rfmurphy81 commented:
@bastiaanv, thanks so much for putting this together. I put this together last night and had a few notes that I was hoping to pass along for you to consider incorporating into the next commit:

1. On the lock screen, the lime green color of the current glucose reading is difficult to read on my light blue background during the daytime (using the "weather" background). See attached. The contrast needs to be sharper. If it were black or a darker color, it would be easier to read. Is this something that is fixable on your end?
![IMG_0482](https://github.com/user-attachments/assets/f8a89984-23f5-4a3c-826f-0d3a85a61673)
2. Also on the lock screen, I'd like to see the "eventual glucose" number somewhere as I use that reference.
3. Also on the lock screen, is it possible to incorporate the basal display somewhere? I'd like to know whether I am getting my scheduled basal or if Loop has modified it. 
4. In the Settings > Notifications > Live activity screen, the Save button doesn't actually do anything so I am unable to actually change any of the default values on this screen. 
I built this last night with the code.

And for those who are not as technically savvy as me, I followed the instructions at https://www.loopandlearn.org/custom-code/ which easily allowed me to pull in a set list of customizations, including this one. :)

bastiaanv commented:
@rfmurphy81 First of all, thank you for your feedback! Really appreciate it.

I will try to find some time to fix the save button soon!

About point 1: The only option currently is to disable BG colouring, but since the save button doesnt work correctly it is hard to try that option. Other idea: would less opacity in the background be a solution for you? Because I think changing the color would be a solution for your background color, but not for someone else. Having less opacity in the widget would leave the widget less vulnerable to this issue.

About point 2: I found the code for the y-axis calculation for the Loop chart, will use that code as soon as I try to fix the save button!

bastiaanv committed:
fix: Fix save with routing back and forth

bastiaanv committed:
style: Removed auto scaling chart

bastiaanv committed:
optimize

bastiaanv commented:
@marionbarker I updated the LA with the following:
- Fix for the Save button
- Removed the auto-scaling and ported the Loop yAxis generator

Do you have time for a update of the customisation?

bastiaanv committed:
hotfix

bastiaanv committed:
hotfix: yAxis scale

marionbarker commented:
The most recent updates to Live Activity have been incorporated in the customization script. Run the script again and you will be offered an update to Live Activity if you already have it installed.

```
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/loopandlearn/lnl-scripts/main/CustomizationSelect.sh)"
```

Documentation overview for the script: https://www.loopandlearn.org/custom-code/#overview

cachrisman commented:
Just wanted to share something I just noticed - that with the live activity running, if my phone is in StandBy (mine is on a MagSafe desk charger, it shows my BG as a small bubble in the upper right of the clock and I can tap that and see the live activity full screen. Pretty neat!!

Keep up the good work everyone! Thank you for all you do!

![IMG_0108](https://github.com/user-attachments/assets/6e97bfe7-b9b4-425d-9fdf-c244f831d29e)
![IMG_0109](https://github.com/user-attachments/assets/a4945af9-4c3c-423b-a3d6-1275f0d9d0c6)

bastiaanv committed:
feat: Allow LA to be hidden when nothing is happening or to be always on

bastiaanv committed:
Merge branch 'LoopKit:dev' into feat/live-activity

bastiaanv commented:
## Progress update:
Allow users to always show the LA or to show the live activity on specific triggers (for now: High or low prediction)

| Settings                                                                                     | Live Activity trigger warning                                                                |
|----------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------|
| ![IMG_0031](https://github.com/user-attachments/assets/55dabba9-299b-4bac-8e0c-cd40bfc25e50) | ![IMG_0030](https://github.com/user-attachments/assets/113bcd8a-90e0-493a-9c4a-b29d2c8da261) |

marionbarker commented:
Sorry to be slow testing this.
I configured it on an SE.
I modified the Live Activity setting to disable the Always Enabled, and disabled the Show when Low is predicted, enabled the Show when High only.
However, the LA on Lock screen still showed up (there was a low at the time.)

Also, IMO, the Save button should be inactive unless there is a change to save. Then when it is tapped to save a change, the Save button should be inactive again.

marionbarker commented:
I did more testing - this is with an SE so I can only look at the lock screen - there is no dynamic island. 

When `Always enabled` is off and glucose is between low and high settings, the lock screen shows the note on the left of the graphic below. (Tapping to update the widget does nothing unless you modify the setting to enable the `Always enabled` slider.)

Also, the two sliders for `Show when low is predicted` and `Show when high is predicted` have no effect. The live activity on the lock screen shows when glucose is below low or above high values regardless of whether those sliders are selected or not.

![Live_Activity_pr2191](https://github.com/user-attachments/assets/1a27e05f-2353-4531-89be-884531b165ad)

bastiaanv committed:
revert: Always on feature

bastiaanv committed:
chore: Save button only enabled when state is dirty

bastiaanv commented:
Reverted always on feature, since Apple doesn't allow a live activity to be created in the background. Only in the foreground or via push notifications.

Also added the change to have the save button disabled while there are no changes

marionbarker commented:
## Summary:

* I have the latest version of this PR on my Looping phone. (commit `cc8ee11`).
* I have been using it since last summer.
* I strongly recommend this be added to Loop as a feature; but there are some issues that could be addressed now or later.
* This feature is enabled by entering Loop, Alert Management and using the new row labeled Live Activity.

## Tests:

I have not tested this in a while. It seems to have the same issue we see in other features where you make a change, and the change is actually implemented, but it doesn't show up on the view. By going to another screen (or quit/restart the app) you can force the refresh for that screen with current values.

This happens for some Therapy Settings and Algorithm Experiment selections as well. It appears to be new with iOS 18.

This is an Issue reported for one display glitch and a PR proposed for another one:
* Issue #2267 
* PR #2279 

I think we need a uniform fix for this for iOS 18.

### Screens can show old settings

I did some testing with the screen where you can choose which 4 blocks you want on the bottom row on the lockscreen. The changes showed up on the Live Activity, but if I returned to the Bottom Row design in Loop, they showed the old selection. A quit/restart updated the view so my current selections were shown.

### Save Button always active

The save button is still always active whether a change was made or not.

### Features worked

Except for the display glitches mention above, all features and changes worked as expected.

### Graphics of Controls

Some of the controls are shown on the screenshots below.

![pr-2191-controls](https://github.com/user-attachments/assets/30bd9b52-354c-4fba-98a8-59eb18602291)

### Graphics on Lockscreen

The screenshots below show the lock screen
* left side is when Mode: Small is selected (only shows the blocks selected for the Bottom Row)
* right side when Mode: Large  is selected with prediction lines enabled.

![pr-2191-lockscreen](https://github.com/user-attachments/assets/3143ab4b-6a3a-47c1-8b27-7b9511ad00e9)

marionbarker reviewed (approved):
I approve this PR, but acknowledge it is subject to some UI glitches on some of the control views.

In practice, most people will configure the Live Activity layout one time and leave it alone. I have not modified the layout for mine since I last tested this PR.

However, if the UI glitches can be fixed easily, they should be.

mns6754 commented:
Hi,
I didn’t build my own loop app, installed it through my health care provider (BC Diabetes). I’m not really a tech savvy; however, have a good handle on the loop app features (I think). I’m using iPhone 12 mini. I couldn’t find the Live Activity option in Alert Management. Is there any way to get this feature on my loop app/ phone even if I haven’t built my own loop app? After upgrading to IOS 18, the loop widget has become useless.

bastiaanv commented:
Hi @mns6754 ,

I am afraid not. Your health care provider needs to changes their Loop build to include LA.

Docs: https://www.loopandlearn.org/loop-features-in-development/#live-activity

mns6754 commented:
> Hi @mns6754 ,
> 
> I am afraid not. Your health care provider needs to changes their Loop build to include LA.
> 
> Docs: https://www.loopandlearn.org/loop-features-in-development/#live-activity

Thank you, Bastiaan, for the clarification

bastiaanv committed:
fix: small configuration fixes

bastiaanv committed:
Merge branch 'LoopKit:dev' into feat/live-activity

marionbarker commented:
## Status

Live Activity controls work as expected.
LGTM.

## Test

Test again with the latest updates. Focus on the whether the configuration displays the just-selected selections after a change. (Before they were enacted, but the screen did not update right away.)

### Configuration

IPhone SE 3rd gen, running iOS 18.3.2
* LoopWorkspace dev, commit `3d48a5c`
* update Loop branch from this PR plus commit from PR 2298 (fix error accidentally introduced with PR 2295)

### Test Save Button and Configuration Display

The Save button acts appropriately again (see this [comment](https://github.com/LoopKit/Loop/pull/2191#issuecomment-2741968011) for details). This behavior is related to iOS 18 - previous tests were with iOS 17.

Initially, the Live Activity screen Save button was inactive
* Modify the Live Activity Mode from Large to Small, Save button becomes active
* Tap save and save button becomes inactive
* Tap back and return, Mode of Small is correctly displayed
* Modify the Live Activity Mode from Small to Large, Save button becomes active
* Tap back (without saving) and then return
   * Success: Mode still shows as Small and the Save button is inactive

### Comprehensive Tests for Save Button and Configuration Display

Tests below include looking at the Live Activity on the Lock Screen. For "success", the LA must match the displayed configuration.

Repeat test for all rows on the Live Activity screen:
* success: Enabled
* success: Mode
* success: Add predictive line (Large mode only)
* success: Use BG coloring
* success: Upper limit
* success: Lower limit

### Test Bottom Row configuration screen

Tests below include looking at the Live Activity on the Lock Screen. For "success", the LA must match the displayed configuration.
* success: delete one item, save button active - back and return, no change
* success: delete one item, save button active- save, back and return, change is displayed on screen
* success: add one item, save button active - back and return, no change
* success: add one item, save button active- save, back and return, change is displayed on screen
* success: rearrange order, save button active - back and return, no change
* success: rearrange order, save button active- save, back and return, change is displayed on screen

marionbarker commented on Loop.xcodeproj/project.pbxproj:
```diff
@@ -4832,6 +4921,7 @@
 				INFOPLIST_FILE = "Loop Widget Extension/Bootstrap/Info.plist";
 				INFOPLIST_KEY_CFBundleDisplayName = "Loop Widgets";
 				INFOPLIST_KEY_NSHumanReadableCopyright = "Copyright © 2022 LoopKit Authors. All rights reserved.";
+				IPHONEOS_DEPLOYMENT_TARGET = 16.2;
```
Please remove this change (found 8 times) from the pbxproj file.
Instead update `Loop.xcconfig` and change this section:

from:
```
// Min iOS Version [DEFAULT]                                    
IPHONEOS_DEPLOYMENT_TARGET = 15.1
```

to:
```
// Min iOS Version [DEFAULT]                                    
IPHONEOS_DEPLOYMENT_TARGET = 16.2
```

marionbarker reviewed (changes requested):
I commented on the need to use Loop.xcconfig to change IPHONEOS_DEPLOYMENT_TARGET. Withdrawing approval until that is updated.

bastiaanv committed:
fix: update IPHONEOS_DEPLOYMENT_TARGET

bastiaanv committed:
Merge branch 'feat/live-activity' of https://github.com/bastiaanv/Loop into HEAD

bastiaanv commented:
@marionbarker Change request is pushed!

marionbarker commented:
## Resolution Needed

This code does not build as is because of 2 problems.

Once I resolved both of these in my local copy of your feat/live-activity branch, the code built and live-activity is displayed on my test phone (SE 3rd generation).

1. I made a mistake and merged a fix meant for tidepool-merge branch into dev for the Bolus Progress display glitch. I later fixed it with another PR that merged to dev at a later time - you picked up the first half of the fix (PR 2295) and not the second when you merged dev into your branch. Please update with latest Loop dev that includes PR 2298.

2. Turns out a couple of hard coded values for IPHONEOS_DEPLOYMENT_TARGET are in the `Loop.xcodeproj/project.pbxproj` file (blame says about 2 years ago).

This is from dev branch, and those lines are also there in feat/live-activity branch (at lines 5085 and 5195).  
```
Loop % grep -n IPHONEOS_DEPLOYMENT_TARGET Loop.xcodeproj/project.pbxproj
4996:				IPHONEOS_DEPLOYMENT_TARGET = 15.1;
5106:				IPHONEOS_DEPLOYMENT_TARGET = 15.1;
```

The lines with hard-coded values in `Loop.xcodeproj/project.pbxproj` need to be removed for the value of `IPHONEOS_DEPLOYMENT_TARGET ` in `Loop.xcconfig` to be recognized. But since that change came in with a tidepool sync in 2023 [Loop PR 1991](https://github.com/LoopKit/Loop/pull/1991) - let's find out if @ps2 wants a separate PR to remove those, or have that removal come in with this feature branch.

ps2 commented:
It's ok to remove IPHONEOS_DEPLOYMENT_TARGET settings in this PR.

bastiaanv committed:
fix: remove extra IPHONEOS_DEPLOYMENT_TARGET

bastiaanv committed:
Merge branch 'LoopKit:dev' into feat/live-activity

bastiaanv commented:
@marionbarker fixed! branch is up-to-date with dev and removed the extra `IPHONEOS_DEPLOYMENT_TARGET`

marionbarker reviewed (approved):
## Status

All good. LGTM.

## Test

Begin with LoopWorkspace dev clone.
In Loop, switch to this branch and rebuild.
Build works as expected.

When built on a phone that supports Dynamic Island, the glucose shows up in the Dynamic Island when Live Activity is enabled.

The Live Activity, on the Lock Screen, works for
* SE 3rd running iOS 18.3.2
* iPhone 8 running iOS 16.7.10

ps2 commented:
Colors should match the Loop standard colors; glucose should not be orange, for example.  The glucose graph colors should match the normal Loop graph colors if possible. The weight of the glucose points should be adjusted for the more dense display (right now all the green dots run into each other), and the forecast should be thin long dashed lines.

We generally try to avoid jargon like "IOB" and "COB" in the app. I know it might take some redesign to fit it, but we've tried to maintain this, even on the tiny watch interface.

IOB -> Active Insulin
COB -> Active Carbohydrates, or Active Carbs

Thanks!

mns6754 commented:
Thank you, Bastiaan and Marion, for developing and adding the Live Activity feature to the Loop. It’s a great addition to the app. 
Personally, everything is good for what I need from this feature, but I thought adding the pump status (reservoir fill status) could be useful, as graphic or in digits format next to the glucose graph on the top section, and/or to the list of “Bottom Row Configuration” options. I think for people like me who use only three items in the bottom row to get a full reading of the Current Glucose number (and to keep the bottom row tidy to quickly see what is important, and with just a glance), the preferred location would be in top section next to the glucose graph, if possible. Thanks,

mns6754 commented:
In reference to my comment above, these pictures are for comparing the options of having three or four items in the bottom row. Personally, I find three items to be optimal in this current view design. 

![IMG_7306](https://github.com/user-attachments/assets/1d2a632a-5ebf-4ea8-a83e-73cc44f22bcf)
![IMG_7331](https://github.com/user-attachments/assets/fda61e9f-0531-48e5-b0c5-f11959273b45)

marionbarker commented:
Reporting a display glitch when using 24 hour clock. Note the time stamp (for 13:00), shows a 1.

![pr2191-display-glitch-24-hour](https://github.com/user-attachments/assets/c82960ad-9d9d-4bd9-a78a-2c8e47a3d138)

bastiaanv committed:
fix: update label & label size

bastiaanv committed:
Merge branch 'feat/live-activity' of https://github.com/bastiaanv/Loop into HEAD

elnjensen commented:
Hi @ps2 - A question as I'm working on this: 

> Colors should match the Loop standard colors; glucose should not be orange, for example. The glucose graph colors should match the normal Loop graph colors if possible.

Do you object to different glucose colors being an *option*, but not the default?  I've updated the code to use the standard Loop blue for the glucose points and prediction line (which also now match the main graph style better), but Bastiaan's code exposes a user option for color-coding the glucose points based on whether they are high, low, or in-range. 

A possible middle ground would be to retain that color coding only for the display of the current BG number below the graph, but not the graph points themselves.  That could be handy for glanceability, including in the CarPlay display when driving, and on the Watch Smart Stack.  There's no graph in those views, but seeing the color of the BG number out of the corner of your eye while driving would tell you whether or not you're in range, even if you don't read the number itself. 

I'm working on the other issues you flagged, and hope to share something soon.  @bastiaanv, is it simplest for me to PR possible changes against your branch?  Then if you merge, changes would automatically feed into this PR?   Or let me know if another route is preferable.

elnjensen committed:
Change chart styling to be more like main app

elnjensen committed:
Add small view for CarPlay and Watch Smart Stack

bastiaanv commented:
@elnjensen A PR against my branch would be the simplest flow yes. I will resolve any merge conflicts that are currently present on this PR afterwards

elnjensen committed:
Add delta BG below main row

bastiaanv committed:
Merge pull request #4 from elnjensen/live-activity-edits

Live activity edits

bastiaanv committed:
merge dev

bastiaanv committed:
fix: small merge conflict fixes

bastiaanv committed:
feat: small WatchOS & carplay change

bastiaanv commented:
Update by @elnjensen!

| Lock-screen                                                                                                              | WatchOS & Carplay                                                                                                                                    |
|--------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------|
| <img width="250" alt="IMG_2678" src="https://github.com/user-attachments/assets/8c454cb9-08de-45d5-9370-7aed17da0f1d" /> | <img width="400" alt="Main-VideoStream-2025-11-01-17-36-49" src="https://github.com/user-attachments/assets/313ca78a-79f2-47e6-9a2e-41f9e7fd8203" /> |


Changes:
- Synced Loop dev to this PR
- Processed feedback from @ps2  ([link](https://github.com/LoopKit/Loop/pull/2191#issuecomment-2968807275))
- Added custom view for WatchOS & Carplay

marionbarker commented:
Thanks for the updates @elnjensen and @bastiaanv.
LGTM

## Test

This feature has been used as part of a customization for more than a year.
The requests from @ps2 were addressed.
I did not test with CarPlay, but I did test on my personal phone with my watch.

**NOTE: the new iOS minimum supported is iOS 16.**

### Test details

started with LoopWorkspace `update_to_dev_3.9.3` branch to see if we can add this feature to that update.
locally merged Loop `dev` branch into the `feat/live-activities` commit 4310a3bc
Built with no issues onto these physical phones:
* SE 2nd gen phone running iOS 18.7.2
* SE 3rd gen phone running iOS 26.1
* iPhone 15 running iOS 18.7.2
* iPhone 8 running iOS 16.7.12

The graphic below shows
* left side is the version before @elnjensen made his updates
* right side shows the updated display

<img width="2420" height="572" alt="pr-2191-updates" src="https://github.com/user-attachments/assets/daf04793-0fc9-44b8-806c-76b96fb0b3f6" />

The graphic below shows
* left side is the default BG coloring
* right side shows `Use BG coloring` enabled with the high limit set to 120 mg/dL
   * the color is displayed in the dynamic island and on the watch 

<img width="2402" height="561" alt="pr-2191-bg-color" src="https://github.com/user-attachments/assets/40f073ac-a729-405e-b221-32ad3fd26c2a" />
